File: src/com/ssynhtn/mock/PlusOne.java
package com.ssynhtn.mock;

public class PlusOne {
    public int[] plusOne(int[] digits) {
        int c = 1;
        int n = digits.length;
        for (int i = n-1; i>= 0; i--) {
            int sum = c + digits[i];
            digits[i] = sum % 10;
            c = sum >= 10 ? 1 : 0;
        }

        if (c == 0) return digits;
        int[] res = new int[n+1];
        res[0] = c;
        System.arraycopy(digits, 0, res, 1, n);
        return res;
    }
}

File: src/com/ssynhtn/medium/PlusOne.kt
package com.ssynhtn.medium

class PlusOne {
    fun plusOne(digits: IntArray): IntArray {
        var add = 1
        for (i in digits.size - 1 downTo 0) {
            var sum = add + digits[i]
            if (sum <= 9) {
                digits[i] = sum
                return digits
            }

            digits[i] = sum - 10
            add = 1
        }

        val result = IntArray(digits.size + 1)
        result[0] = 1
        System.arraycopy(digits, 0, result, 1, digits.size)
        return result
    }
}

fun main() {
    println(PlusOne().plusOne(intArrayOf(9, 9)).toList())
}